akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpEntity.scala

```diff
@@ -398,8 +398,10 @@ object HttpEntity {
     def withContentType(contentType: ContentType): HttpEntity.Default =
       if (contentType == this.contentType) this else copy(contentType = contentType)
 
-    override def withSizeLimit(maxBytes: Long): HttpEntity.Default =
-      copy(data = Limitable.applyForByteStrings(data, SizeLimit(maxBytes, Some(contentLength))))
+    override def withSizeLimit(maxBytes: Long): HttpEntity.Default = {
+      if (data ne Source.empty) copy(data = Limitable.applyForByteStrings(data, SizeLimit(maxBytes, Some(contentLength))))
```

Comment:
Did I say delicate business? E.g. here we have to special case to preserve `Source.empty` if possible.